File: vane-core/src/main/java/org/oddlama/vane/core/config/recipes/CookingRecipeDefinition.java
package org.oddlama.vane.core.config.recipes;

import java.util.HashMap;
import java.util.Map;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.Tag;
import org.bukkit.inventory.BlastingRecipe;
import org.bukkit.inventory.CampfireRecipe;
import org.bukkit.inventory.FurnaceRecipe;
import org.bukkit.inventory.Recipe;
import org.bukkit.inventory.SmokingRecipe;
import org.oddlama.vane.util.ItemUtil;

public class CookingRecipeDefinition extends RecipeDefinition {

    private String input = null;
    private String result = null;
    private float experience = 0.0f;
    private int cookingTime = 10;
    private String type;

    public CookingRecipeDefinition(String name, String type) {
        super(name);
        this.type = type;
        switch (this.type) {
            case "blasting": // fallthrough
            case "furnace": // fallthrough
            case "campfire": // fallthrough
            case "smoking":
                break;
            default:
                throw new IllegalArgumentException("Invalid cooking recipe: Unknown type '" + this.type + "'");
        }
    }

    public CookingRecipeDefinition input(String input) {
        this.input = input;
        return this;
    }

    public CookingRecipeDefinition input(final Tag<?> tag) {
        return input("#" + tag.key());
    }

    public CookingRecipeDefinition input(Material material) {
        return input(material.key().toString());
    }

    public CookingRecipeDefinition result(String result) {
        this.result = result;
        return this;
    }

    @Override
    public Object toDict() {
        final HashMap<String, Object> dict = new HashMap<>();
        dict.put("CookingTime", this.cookingTime);
        dict.put("Experience", this.experience);
        dict.put("Input", this.input);
        dict.put("Result", this.result);
        dict.put("Type", type);
        return dict;
    }

    @Override
    public RecipeDefinition fromDict(Object dict) {
        if (!(dict instanceof Map<?, ?>)) {
            throw new IllegalArgumentException(
                "Invalid " + type + " recipe dictionary: Argument must be a Map<String, Object>!"
            );
        }
        final var dictMap = (Map<?, ?>) dict;
        final Object inputObj = dictMap.containsKey("Input") ? dictMap.get("Input") : dictMap.get("input");
        if (inputObj instanceof String input) {
            this.input = input;
        } else {
            throw new IllegalArgumentException("Invalid " + type + " recipe dictionary: input must be a string");
        }

        final Object resultObj = dictMap.containsKey("Result") ? dictMap.get("Result") : dictMap.get("result");
        if (resultObj instanceof String result) {
            this.result = result;
        } else {
            throw new IllegalArgumentException("Invalid " + type + " recipe dictionary: result must be a string");
        }

        final Object experienceObj = dictMap.containsKey("Experience") ? dictMap.get("Experience") : dictMap.get("experience");
        if (experienceObj instanceof Float experience) {
            this.experience = experience;
        } else {
            throw new IllegalArgumentException("Invalid " + type + " recipe dictionary: experience must be a float");
        }

        final Object cookingTimeObj = dictMap.containsKey("CookingTime") ? dictMap.get("CookingTime") : dictMap.get("cookingTime");
        if (cookingTimeObj instanceof Integer cookingTime) {
            this.cookingTime = cookingTime;
        } else {
            throw new IllegalArgumentException("Invalid " + type + " recipe dictionary: cookingTime must be a int");
        }

        return this;
    }

    @Override
    public Recipe toRecipe(NamespacedKey baseKey) {
        final var out = ItemUtil.itemstackFromString(this.result).getLeft();
        final var in = RecipeDefinition.recipeChoice(input);
        switch (this.type) {
            case "blasting":
                return new BlastingRecipe(key(baseKey), out, in, experience, cookingTime);
            case "furnace":
                return new FurnaceRecipe(key(baseKey), out, in, experience, cookingTime);
            case "campfire":
                return new CampfireRecipe(key(baseKey), out, in, experience, cookingTime);
            case "smoking":
                return new SmokingRecipe(key(baseKey), out, in, experience, cookingTime);
            default:
                throw new IllegalArgumentException("Invalid cooking recipe: Unknown type '" + this.type + "'");
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/recipes/CookingRecipeDefinition.kt
package org.oddlama.vane.core.config.recipes

import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.Tag
import org.bukkit.inventory.*
import org.oddlama.vane.util.ItemUtil

/**
 * Recipe definition for furnace-like cooking recipes.
 *
 * @param name logical recipe name.
 * @param type cooking recipe type (`blasting`, `furnace`, `campfire`, `smoking`).
 */
class CookingRecipeDefinition(name: String?, private val type: String) : RecipeDefinition(name) {
    /** Input ingredient definition. */
    private var input: String? = null
    /** Result item definition. */
    private var result: String? = null
    /** Experience awarded by the recipe. */
    private var experience = 0.0f
    /** Cooking time in ticks. */
    private var cookingTime = 10

    init {
        when (this.type) {
            "blasting", "furnace", "campfire", "smoking" -> {}
            else -> throw IllegalArgumentException("Invalid cooking recipe: Unknown type '" + this.type + "'")
        }
    }

    /** Sets input ingredient definition. */
    fun input(input: String?): CookingRecipeDefinition {
        this.input = input
        return this
    }

    /** Sets input ingredient by Bukkit tag. */
    fun input(tag: Tag<*>): CookingRecipeDefinition {
        return input("#" + tag.key())
    }

    /** Sets input ingredient by material key. */
    fun input(material: Material): CookingRecipeDefinition {
        return input(material.key().toString())
    }

    /** Sets result item definition. */
    fun result(result: String?): CookingRecipeDefinition {
        this.result = result
        return this
    }

    /** Serializes this recipe definition to dictionary form. */
    override fun toDict(): Any {
        val dict = mutableMapOf<String?, Any?>()
        dict["CookingTime"] = this.cookingTime
        dict["Experience"] = this.experience
        dict["Input"] = this.input
        dict["Result"] = this.result
        dict["Type"] = type
        return dict
    }

    /** Loads this recipe definition from dictionary form. */
    override fun fromDict(dict: Any?): RecipeDefinition {
        require(dict is MutableMap<*, *>) { "Invalid $type recipe dictionary: Argument must be a Map<String, Object>!" }
        val inputObj = if (dict.containsKey("Input")) dict["Input"] else dict["input"]
        if (inputObj is String) {
            this.input = inputObj
        } else {
            throw IllegalArgumentException("Invalid $type recipe dictionary: input must be a string")
        }

        val resultObj = if (dict.containsKey("Result")) dict["Result"] else dict["result"]
        if (resultObj is String) {
            this.result = resultObj
        } else {
            throw IllegalArgumentException("Invalid $type recipe dictionary: result must be a string")
        }

        val experienceObj =
            if (dict.containsKey("Experience")) dict["Experience"] else dict["experience"]
        if (experienceObj is Float) {
            this.experience = experienceObj
        } else {
            throw IllegalArgumentException("Invalid $type recipe dictionary: experience must be a float")
        }

        val cookingTimeObj =
            if (dict.containsKey("CookingTime")) dict["CookingTime"] else dict["cookingTime"]
        if (cookingTimeObj is Int) {
            this.cookingTime = cookingTimeObj
        } else {
            throw IllegalArgumentException("Invalid $type recipe dictionary: cookingTime must be a int")
        }

        return this
    }

    /** Converts this definition into a concrete Bukkit cooking recipe. */
    override fun toRecipe(baseKey: NamespacedKey?): Recipe? {
        val bk = baseKey ?: throw IllegalArgumentException("baseKey cannot be null")
        val out = ItemUtil.itemstackFromString(this.result!!).getLeft()!!
        val `in` = recipeChoice(input!!)
        return when (this.type) {
            "blasting" -> BlastingRecipe(key(bk), out, `in`, experience, cookingTime)
            "furnace" -> FurnaceRecipe(key(bk), out, `in`, experience, cookingTime)
            "campfire" -> CampfireRecipe(key(bk), out, `in`, experience, cookingTime)
            "smoking" -> SmokingRecipe(key(bk), out, `in`, experience, cookingTime)
            else -> throw IllegalArgumentException("Invalid cooking recipe: Unknown type '" + this.type + "'")
        }
    }
}
